helix-vcs/src/differ.rs

```diff
@@ -0,0 +1,164 @@
+use std::mem::take;
+use std::ops::{Deref, Range};
+use std::sync::Arc;
+
+use arc_swap::ArcSwap;
+use helix_core::{Rope, RopeSlice};
+use imara_diff::intern::InternedInput;
+use imara_diff::Algorithm;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::JoinHandle;
+use tokio::time::{timeout, Duration};
+
+use crate::rope_line_cache::InternedRopeLines;
+use crate::{LineDiff, LineDiffs};
+
+#[cfg(test)]
+mod test;
+
+#[derive(Clone, Debug)]
+pub struct Differ {
+    channel: UnboundedSender<Event>,
+    line_diffs: Arc<ArcSwap<LineDiffs>>,
+}
+
+impl Differ {
+    pub fn new(diff_base: Rope, doc: Rope) -> Differ {
+        Differ::new_with_handle(diff_base, doc).0
+    }
+
+    fn new_with_handle(diff_base: Rope, doc: Rope) -> (Differ, JoinHandle<()>) {
+        let (sender, receiver) = unbounded_channel();
+        let line_diffs: Arc<ArcSwap<LineDiffs>> = Arc::default();
+        let worker = DiffWorker {
+            channel: receiver,
+            line_diffs: line_diffs.clone(),
+            new_line_diffs: LineDiffs::default(),
+        };
+        let handle = tokio::spawn(worker.run(diff_base, doc));
+        let differ = Differ {
+            channel: sender,
+            line_diffs,
+        };
+        (differ, handle)
+    }
+    pub fn get_line_diffs(&self) -> impl Deref<Target = impl Deref<Target = LineDiffs>> {
+        self.line_diffs.load()
+    }
+
+    pub fn update_document(&self, doc: Rope) -> bool {
+        self.channel.send(Event::UpdateDocument(doc)).is_ok()
+    }
+
+    pub fn update_diff_base(&self, diff_base: Rope) -> bool {
+        self.channel.send(Event::UpdateDiffBase(diff_base)).is_ok()
+    }
+}
+
+// TODO configuration
+const DIFF_MAX_DEBOUNCE: u32 = 10;
```

Comment:
Should this be 100ms instead? Since 300ms is the time when changes are noticeable to people.

Reply:
good point, your comment made me simplify the debouncer a bit. We now simply collect events for a fixed timeout (which i have change to 100ms)